### DISCUSSION TITLE
Sparsity pattern for Jacobian and Hessian matrices

### DISCUSSION THREAD

**Discussion by a-jp** [vgvassilev/clad#780] [Q&A]
Hi,

I have not used the library yet, but I'm strongly considering porting my code which uses [CppAD ](https://coin-or.github.io/CppAD/doc/cppad.htm) to use Clad.  With CppAD it is possible to obtain sparsity patterns when computing Jacobian and Hessian matrices - entries which are truly zero.  Is this possible with Clad and if so can you point me to any documentation or preferably examples?

Thanks

**Comment by a-jp** [comment]
Or...are the jacobians and hessians that are returned by clad, done so only containing non-zero entries? They are already sparse? If so, how does one infer the location of the zeros when rebuilding a dense matrix?

Thanks

**Comment by vgvassilev** [comment]
Hi @a-jp, thanks for your question.

Currently the way we model jacobians and hessians is without taking into account sparsity. We have an issue which we will tackle soon in that respect #509. Perhaps you could help us understand your usecase by showing a small example?

**Reply by a-jp** [reply]
Hi, Thanks for the reply.  I do not have a small example I could post, my working codebase is quite large and complex.  However, the area of usage is well known.  I would like to use clad to help me provide derivative, gradient, jacobian and hessian information to **non-linear** optimisation frameworks.  They all normally require this info - I'm ignoring optimisation approaches that approximate the hessian.  The optimisation framework I'm using makes significant usage of the sparsity pattern of the jacobian and hessian matrices and while it will allow a dense matrix, significant savings are to be had when you use the sparsity pattern.  This is all well tried using CppAD and works very well.  Clad would allow me to compute the info I need once, at compile time, rather than during run time.  My use case means I call my code in a loop, and not only am I paying the cost of the run time computation of CppAD, but I'm paying the cost of the constant recalculation within the the loop.  There are also some subtilties with retaping that I wish to move past when using CppAD.

Sparsity info is provided to the optimisation framework I use at construction time of the solver, once, so it's a one off cost.  I should be clear about zeros numerically and genuine zero (as would be obtained with book work).  What you need to promise to the optimisation framework are the zeros that will never have a non-zero value, not just zeros because the initial matrix with certain choices of initial constants or indeed catastrophic cancellation give a zero at the point at which you compute the sparsity pattern - in the optimisation framework I'm using, you must know the sparsity pattern before the optimisation is undertaken so in this instance you really need to compute the sparsity as if you had used a symbolic algebra package and not ascribed any values to the dependent, independent and constant variables.

I think obtaining current hessians and jacobians, and then a second call to provide sparsity would be totally fine, this is not something that should need to be recomputed.  We just need to be clear that the zeros returned in the sparsity check, are assured to be zero first or second derivatives, and this has not be obtained assuming any actual value of constants terms. 

Thanks for the link to the other issue, if I understand correctly that relates to diagonal entries, my requirement is quite general and the zero entries could be anywhere.

**Reply by vgvassilev** [reply]
I think I can picture your use-case a little better now. You are coming at the right time because that area is work in progress in the next couple of months. 

> I think obtaining current hessians and jacobians, and then a second call to provide sparsity would be totally fine, this is not something that should need to be recomputed. We just need to be clear that the zeros returned in the sparsity check, are assured to be zero first or second derivatives, and this has not be obtained assuming any actual value of constants terms.

What's a good interface to express this general sparsity in the hessians? IIRC, currently one can express the desirable hessian "by hand" by calling `clad::differentiate` and subsequently `clad::gradient` expressing the necessary directions. Is that something that we could start from?

**Reply by a-jp** [reply]
> What's a good interface to express this general sparsity in the hessians? IIRC, currently one can express the desirable hessian "by hand" by calling clad::differentiate and subsequently clad::gradient expressing the necessary directions. Is that something that we could start from?

Hi thanks for the reply. I'm not sure I follow this question?  Could you elaborate?  

I may have inadvertently caused confusion with my own comment.  I was indicating that there would be no reason to break the api, we could have `clad::hessian_with_sparsity` or something similar, or indeed a call to `clad::hessian` and then if the underlying mechanics stored the sparsity elsewhere(?), a second call to `clad::sparsity(hess)` to obtain the sparsity.  I think we could make the interface quite clear that as it stands all calls currently return dense matrices, and additional semantics are required for sparse versions. 

As to the form of the sparsity information, I personally really don't like the compressed storage format unless it's very well documented, it's a great way to make a mistake unless documented well.  I know I've already said this, but, borrowing from previous compile time expression template code I've written to compute derivatives, you always need a Zero termination type (along with a One type and Constant type etc) - in that analogy we really only want to indicate something is sparse if it had terminated at a Zero type.

As per the above, could you elaborate on your question/comment, I just want to be sure I'm not misunderstanding the point you were making.

**Reply by a-jp** [reply]
Hi @vgvassilev I was just following up my most recent post to see if you had any comments? Thanks

**Reply by vgvassilev** [reply]
@a-jp, apologies for the delay. I think I understand your use-case by the interfaces you described a little better now. I think we can tackle your requirement in clad v1.6 which is scheduled for end of May. Do you think you can capture the gist of this discussion in an issue?

**Reply by a-jp** [reply]
Hi, yep I'll take a first stab at that as soon as possible.

**Reply by a-jp** [reply]
Please see issue #803.  Thanks

**Reply by vgvassilev** [reply]
Thank you!
